org-code-javabuilder/lib/src/main/java/org/code/javabuilder/CodeExecutionManager.java

```diff
@@ -0,0 +1,145 @@
+
+import java.io.File;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.PrintStream;
+import java.nio.file.Files;
+import java.util.List;
+import java.util.logging.Logger;
+import org.code.protocol.*;
+import org.code.protocol.LoggerUtils.ClearStatus;
+import org.code.protocol.LoggerUtils.SessionTime;
+
+/**
+ * Manages the execution of code in a project. When started, it performs the necessary pre-execution
+ * setup and starts code execution. When complete, or when asked to exit early, it ensures the
+ * necessary post-execution cleanup steps are performed.
+ */
+public class CodeExecutionManager {
+  private final ProjectFileLoader fileLoader;
+  private final InputHandler inputHandler;
+  private final OutputAdapter outputAdapter;
+  private final ExecutionType executionType;
+  private final List<String> compileList;
+  private final JavabuilderFileManager fileManager;
+  private final LifecycleNotifier lifecycleNotifier;
+
+  private File tempFolder;
+  private InputRedirectionStream overrideInputStream;
+  private OutputPrintStream overrideOutputStream;
+  private InputStream systemInputStream;
+  private PrintStream systemOutputStream;
+  private boolean executionInProgress;
+
+  public CodeExecutionManager(
+      ProjectFileLoader fileLoader,
+      InputHandler inputHandler,
+      OutputAdapter outputAdapter,
+      ExecutionType executionType,
+      List<String> compileList,
+      JavabuilderFileManager fileManager,
+      LifecycleNotifier lifecycleNotifier) {
+    this.fileLoader = fileLoader;
+    this.inputHandler = inputHandler;
+    this.outputAdapter = outputAdapter;
+    this.executionType = executionType;
+    this.compileList = compileList;
+    this.fileManager = fileManager;
+    this.lifecycleNotifier = lifecycleNotifier;
+    this.executionInProgress = false;
+  }
+
+  /** Executes code synchronously, ensuring that pre- and post-execution steps are performed. */
+  public void execute() {
+    try {
+      this.onPreExecute();
+      try {
+        // Run code builder
+        this.executionInProgress = true;
+        final CodeBuilderRunnable runnable =
+            new CodeBuilderRunnable(
+                this.fileLoader,
+                this.outputAdapter,
+                this.tempFolder,
+                this.executionType,
+                this.compileList);
+        runnable.run();
+      } catch (Throwable e) {
+        // Catch any throwable here to ensure that onPostExecute() is always called
+        LoggerUtils.logException(e);
+      }
+      this.onPostExecute();
+    } catch (InternalServerError e) {
+      LoggerUtils.logError(e);
+    }
+  }
+
+  /**
+   * Request the code execution to cleanup early (e.g. in the case of an impending timeout). Code
+   * execution may still run, but this allows for post execution steps to still occur before the
+   * environment is shut down.
+   */
+  public void requestEarlyExit() {
+    if (!this.executionInProgress) {
```

Comment:
Yeah this is just a safeguard because `requestEarlyExit` is a public method - if we somehow accidentally call this from elsewhere before we call `execute()` this ensures that we don't do post execution steps. Very unlikely scenario, but just to be safe :)